networkx/algorithms/approximation/broadcasting.py

```diff
@@ -0,0 +1,98 @@
+import networkx as nx
+from networkx.utils import not_implemented_for
+
+__all__ = [
+    "tree_broadcast",
+]
+
+
+def update_broadcast_label(G, U, v):
+    adj = [u for u in G.neighbors(v) if u in U]
+    adj.sort(key=lambda u: G.nodes[u]["value"], reverse=True)
+    return max([G.nodes[u]["value"] + i for i, u in enumerate(adj, start=1)])
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def tree_broadcast(G):
+    """
+    This functions implements a linear algorithm for determining the minimum broadcast time
+    on any tree [1]_. As a byproduct, it can also find a vertex which acts as the broadcast center,
+    i.e., the vertex where the broadcast begins.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        Undirected graph
+
+    Returns
+    -------
+    b_T : int
+        Optimal Broadcast time of a tree
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If the graph is directed or is a multigraph.
+
+    References
+    ----------
+    .. [1] Slater, P.J., Cockayne, E.J., Hedetniemi, S.T,
+       Information dissemination in trees. SIAM J.Comput. 10(4), 692–701 (1981)
+    """
+    # Remove selfloops if necessary
+    loop_nodes = nx.nodes_with_selfloops(G)
+    try:
+        node = next(loop_nodes)
+    except StopIteration:
+        pass
+    else:
+        G = G.copy()
+        G.remove_edge(node, node)
+        G.remove_edges_from((n, n) for n in loop_nodes)
+
+    # Assert that the graph G has no cycles
+    assert nx.is_tree(G), "The graph G is not a tree"
+
+    # step 0
+    if G.number_of_nodes() == 2:
+        return 1
+    elif G.number_of_nodes() == 1:
+        return 0
+
+    # step 1
+    U = {node for node in G.nodes() if G.degree(node) == 1}
+    for u in U:
+        G.nodes[u]["value"] = 0
```

Comment:
It's probably better/faster/cleaner and (avoids possibly mutating the input graph object) to store the values in a separate dict keyed by node to the value.
```python
values = {n: 0 for n in U}
```

Reply:
ok thanks, good to know for future usages. I did find it quite messy to have ``G.nodes()[u]["values"]``.